src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1248,6 +1220,123 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
             }
         }
 
+        private bool IsRefStruct(out bool isRefFollowedByStructOrPartialStruct, out bool isTypeDeclaration)
+        {
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.RefKeyword);
+
+            var nextToken = PeekToken(1);
+
+            isRefFollowedByStructOrPartialStruct = nextToken.Kind == SyntaxKind.StructKeyword ||
+                (nextToken.ContextualKind == SyntaxKind.PartialKeyword && this.PeekToken(2).Kind == SyntaxKind.StructKeyword);
+
+            if (isRefFollowedByStructOrPartialStruct)
+            {
+                isTypeDeclaration = true;
+                return true;
+            }
+
+            isTypeDeclaration = IsTypeDeclaration(out nextToken);
+            return isTypeDeclaration && nextToken.Kind == SyntaxKind.StructKeyword;
+        }
+
+        private static bool CanFollowTypeName(SyntaxKind kind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.DotToken:                   // partial.     qualified
+                case SyntaxKind.ColonColonToken:            // partial::    alias
+                case SyntaxKind.LessThanToken:              // partial<     generic type
+                case SyntaxKind.OpenBracketToken:           // partial[     array
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static bool CanFollowMemberName(SyntaxKind kind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.LessThanToken:              // partial<     generic method
+                case SyntaxKind.OpenBraceToken:             // partial {    property
+                case SyntaxKind.EqualsGreaterThanToken:     // partial =>   property
+                case SyntaxKind.OpenParenToken:             // partial(     method
+                    return true;
+            }
+
+            return false;
+        }
+
+        private enum PartialModFlags
+        {
+            PartialType,
+            PartialMember,
+            TreatAsModifier,
+            NotModifier
```

Comment:
these would benefit from comments.

Reply:
not sure how they all relate... if you had TreatAsModifier... does that mean we def think it's a modifier... but it wasn't a type or a member?  If so, when does that happens?  Need comments for sure.